### PR TITLE
Fix/fuzzing errors v50

### PR DESCRIPTION
**Pull request by lafirest**
Fixes <issue-number>

**If your build fails** due to your commit message not passing the build checks, please review the guidelines here: https://github.com/emqx/emqx/blob/master/CONTRIBUTING.md.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [x] Changed lines covered in coverage report
- [x] Change log has been added to `changes/` dir
- [x] ~~For EMQX 4.x: `appup` files updated (execute `scripts/update-appup.sh emqx`)~~
- [x] For internal contributor: there is a jira ticket to track this change
      [EMQX-8305](https://emqx.atlassian.net/browse/EMQX-8305)
- [x] ~~If there should be document changes, a PR to emqx-docs.git is sent, or a jira ticket is created to follow up~~
- [x] ~~In case of non-backward compatible changes, reviewer should check this item as a write-off, and add details in **Backward Compatibility** section~~

### CODE REVIEW

File: apps/emqx/src/emqx_frame.erl [hunk 1]
```diff
@@ -182,8 +182,19 @@ parse_remaining_len(
...
 %% Match PINGREQ.
-parse_remaining_len(<<0:8, Rest/binary>>, Header, 1, 0, Options) ->
+parse_remaining_len(
+    <<0:8, Rest/binary>>, Header = #mqtt_packet_header{type = ?PINGREQ}, 1, 0, Options
+) ->
+    parse_frame(Rest, Header, 0, Options);
+parse_remaining_len(
+    <<0:8, Rest/binary>>, Header = #mqtt_packet_header{type = ?PINGRESP}, 1, 0, Options
+) ->
     parse_frame(Rest, Header, 0, Options);
```
Comment (zmstone) [inline apps/emqx/src/emqx_frame.erl:192]: As a server, it should never receive PINGRESP.
We should directly close the connection

File: apps/emqx/src/emqx_frame.erl [hunk 2]
```diff
@@ -182,8 +182,19 @@ parse_remaining_len(
...
+%% Without this clause a crash may will happen when data incorrect, this was found by fuzzing test
+parse_remaining_len(
+    <<0:8, _Rest/binary>>, _Header, 1, 0, _Options
+) ->
+    ?PARSE_ERR(invalid_remaining_len);
```
Comment (zmstone) [inline apps/emqx/src/emqx_frame.erl:197]: maybe raise a map with `#{hint => expecting_non_empty_payload, header_type => Header#mqtt_packet_header.type}` here so we know which

File: apps/emqx/src/emqx_frame.erl [hunk 3]
```diff
@@ -182,8 +182,19 @@ parse_remaining_len(
...
-parse_remaining_len(<<0:8, Rest/binary>>, Header, 1, 0, Options) ->
+parse_remaining_len(
+    <<0:8, Rest/binary>>, Header = #mqtt_packet_header{type = ?PINGREQ}, 1, 0, Options
+) ->
+    parse_frame(Rest, Header, 0, Options);
+parse_remaining_len(
+    <<0:8, Rest/binary>>, Header = #mqtt_packet_header{type = ?PINGRESP}, 1, 0, Options
+) ->
     parse_frame(Rest, Header, 0, Options);
+%% Without this clause a crash may will happen when data incorrect, this was found by fuzzing test
```
Comment (zmstone) [inline apps/emqx/src/emqx_frame.erl:193]:
```suggestion
All other types of messages should not have a zero remaining length.
```

File: apps/emqx/src/emqx_frame.erl [hunk 4]
```diff
@@ -255,20 +266,29 @@ packet(Header, Variable) ->
...
-parse_packet(
-    #mqtt_packet_header{type = ?CONNECT},
-    FrameBin,
-    #{strict_mode := StrictMode}
+parse_connect(FrameBin, StrictMode) ->
+    {ProtoName, Rest} = parse_utf8_string_with_hint(FrameBin, StrictMode, invalid_proto_name),
+    case ProtoName of
+        <<"MQTT">> ->
+            ok;
+        <<"MQIsdp">> ->
```
Comment (zmstone) [inline apps/emqx/src/emqx_frame.erl:274]: lack of regression for this clause.

File: apps/emqx/src/emqx_frame.erl [hunk 5]
```diff
@@ -255,20 +266,29 @@ packet(Header, Variable) ->
...
+            ok;
+        _ ->
+            %% from spec: the server MAY send disconnect with reason code 0x84
+            %% we chose to close socket because the client is likely not talking MQTT anyway
+            ?PARSE_ERR(#{hint => invalid_proto_name, expected => <<"MQTT">>, received => ProtoName})
```
Comment (zmstone) [inline apps/emqx/src/emqx_frame.erl:279]:
```suggestion
            ?PARSE_ERR(#{hint => invalid_proto_name, expected => <<"'MQTT' or 'MQIsdp'">>, received => ProtoName})
```

File: apps/emqx/src/emqx_frame.erl [hunk 6]
```diff
@@ -487,13 +541,13 @@ parse_property(<<16#18, Val:32, Bin/binary>>, Props, StrictMode) ->
 parse_property(<<16#19, Val, Bin/binary>>, Props, StrictMode) ->
     parse_property(Bin, Props#{'Request-Response-Information' => Val}, StrictMode);
 parse_property(<<16#1A, Bin/binary>>, Props, StrictMode) ->
-    {Val, Rest} = parse_utf8_string(Bin, StrictMode),
+    {Val, Rest} = parse_utf8_string_with_hint(Bin, StrictMode, invalid_response_info),
     parse_property(Rest, Props#{'Response-Information' => Val}, StrictMode);
 parse_property(<<16#1C, Bin/binary>>, Props, StrictMode) ->
-    {Val, Rest} = parse_utf8_string(Bin, StrictMode),
+    {Val, Rest} = parse_utf8_string_with_hint(Bin, StrictMode, invalid_sever_ref),
```
Comment (zmstone) [inline apps/emqx/src/emqx_frame.erl:547]:
```suggestion
    {Val, Rest} = parse_utf8_string_with_hint(Bin, StrictMode, invalid_sever_reference),
```

File: apps/emqx/src/emqx_frame.erl (hunk 4 above)
Comment (lafirest) [reply apps/emqx/src/emqx_frame.erl:274]: No, there already were some case to cover this

**Review by zmstone — DISMISSED** [review]

**Review by zmstone — APPROVED** [review]

**Pull request merged**